File: Sliding_Windows/Leetocde_713/code.cpp
#include <iostream>
#include <vector> 

using namespace std;

class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int n = nums.size();
        if (n == 0 || k == 0)
            return 0;
      
        int count = 0;
        for (int i = 0 ; i < n; i++)
        {
            long long prod = 1;
            if (nums[i] < k)
            {
                count++;
                prod *= nums[i];
                for ( int j = i + 1; j < n; j++)
                {
                    prod *= nums[j];
                    if (prod < k) 
                        count++;
                    else 
                        break;
                }
            }
        }
        return count;
    }
};

int main()
{
    vector<int> v = { 10, 5, 2, 6};
    int k = 100;
    Solution sol;
    cout << " No. of contiguous subarrays having prod less than " << k  << " = " 
         << sol.numSubarrayProductLessThanK(v, k) << endl;

    int k1 = 10;
    cout << " No. of contiguous subarrays having prod less than " << k1  << " = " 
         << sol.numSubarrayProductLessThanK(v, k1) << endl;
    return 0;
}

File: Sliding_Windows/Leetocde_713/code.py
from typing import List

class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        l = len(nums)
        if (l == 0 or k == 0):
            return 0
        count = 0
        prod = 1
        left = 0 
        for right, x in enumerate(nums):
            prod *= x
            while (left <= right and prod >= k):
                prod //= nums[left]
                left += 1
            count += right - left + 1
    
        return count
    
if __name__ == "__main__":
    nums = [10,5,2,6]
    K = 10
    K1 = 100
    sol = Solution()
    print (f"No. of Subarrays having prod is less than  {K} is {sol.numSubarrayProductLessThanK(nums, K)}")  # expect 3
    print (f"No. of Subarrays having prod is less than  {K1} is {sol.numSubarrayProductLessThanK(nums, K1)}")  # expect 8
    
